airflow/providers/openlineage/utils/sql.py

```diff
@@ -0,0 +1,195 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import logging
+from collections import defaultdict
+from contextlib import closing
+from enum import IntEnum
+from typing import TYPE_CHECKING, Dict, List, Optional
+
+from attrs import define
+from sqlalchemy import Column, MetaData, Table, and_, union_all
+
+from openlineage.client.facet import SchemaDatasetFacet, SchemaField
+from openlineage.client.run import Dataset
+
+if TYPE_CHECKING:
+    from sqlalchemy.engine import Engine
+    from sqlalchemy.sql import ClauseElement
+
+    from airflow.hooks.base import BaseHook
+
+
+logger = logging.getLogger(__name__)
+
+
+class ColumnIndex(IntEnum):
+    """Enumerates the indices of columns in information schema view."""
+
+    SCHEMA = 0
+    TABLE_NAME = 1
+    COLUMN_NAME = 2
+    ORDINAL_POSITION = 3
+    # Use 'udt_name' which is the underlying type of column
+    UDT_NAME = 4
+    # Database is optional as 5th column
+    DATABASE = 5
+
+
+TablesHierarchy = Dict[Optional[str], Dict[Optional[str], List[str]]]
+
+
+@define
+class TableSchema:
+    """Temporary object used to construct OpenLineage Dataset."""
+
+    table: str
+    schema: str | None
+    database: str | None
+    fields: list[SchemaField]
+
+    def to_dataset(self, namespace: str, database: str | None = None) -> Dataset:
+        # Prefix the table name with database and schema name using
+        # the format: {database_name}.{table_schema}.{table_name}.
+        name = ".".join(
+            [
+                x
+                for x in [self.database if self.database else database, self.schema, self.table]
+                if x is not None
+            ]
+        )
+
+        return Dataset(
+            namespace=namespace,
+            name=name,
+            facets={"schema": SchemaDatasetFacet(fields=self.fields)} if len(self.fields) is not None else {},
```

Comment:
The `len()` part can never be None and thus the `else` part can never execute. Is there a typo here?

Reply:
It is, fixed.